gabbi/case.py

```diff
@@ -402,7 +406,34 @@ def _test_status(self, expected_status, observed_status):
             statii = [stat.strip() for stat in expected_status.split('||')]
         else:
             statii = [expected_status.strip()]
-        self.assertIn(observed_status, statii)
+
+        self.assert_in_or_print_output(observed_status, statii)
+
+    def assert_in_or_print_output(self, expected, iterable):
+        if self.content_type != "application/binary":
+            if self.json_data:
+                full_response = json.dumps(self.json_data, indent=2,
+                                           separators=(',', ': '))
+            else:
+                full_response = self.output
+
+            max_lines = os.getenv('GABBIT_MAX_OUTPUT_LINES', MAX_OUTPUT_LINES)
+            response = "\n".join(full_response.split('\n')[0:max_lines])
+            is_truncated = (len(response) != len(full_response))
+
+            if iterable == self.output:
+                msg = "'%s' not found in %s%s" % (
+                    expected, response,
+                    '\n...truncated...' if is_truncated else ''
+                )
+            else:
+                msg = "'%s' not found in %s, %sresponse:\n%s" % (
+                    expected, iterable,
+                    'truncated ' if is_truncated else '',
+                    response)
+            self.assertIn(expected, iterable, msg)
```

Comment:
I'm probably prematurely optimizing but it seems like perhaps msg shouldn't be created until we know there has been a failure otherwise all that effort to produce a nice message will have gone to waste.